File: src/utils/parser.js
const cheerio = require('cheerio');

function parseResults(rawResults) {
  const seenUrls = new Set();
  const parsed = [];

  rawResults.forEach(item => {
    let content = item.content || '';
    if (content) {
      const $ = cheerio.load(content);
      content = $.text().trim();
    }

    const url = item.url || '';
    if (!url || seenUrls.has(url)) return;
    seenUrls.add(url);

    let published = item.publishedDate || item.pubdate;
    if (published) {
      try {
        // Handle ISO format and Z suffix
        const date = new Date(published.replace('Z', '+00:00'));
        if (!isNaN(date)) {
          published = date.toISOString().split('T')[0];
        } else {
          published = null;
        }
      } catch (e) {
        published = null;
      }
    } else {
      published = null;
    }

    parsed.push({
      title: item.title || 'Untitled',
      url: url,
      snippet: content.slice(0, 500),
      engine: item.engine || 'unknown',
      publishedDate: published
    });
  });

  return parsed;
}

module.exports = parseResults;

File: src/services/parser.py
from lxml import html

def parse_results(raw_results: list) -> list:
    seen_urls = set()
    parsed = []

    for item in raw_results:
        content = item.get("content", "")
        if content:
            tree = html.fromstring(content)
            content = tree.text_content().strip()

        url = item.get("url", "")
        if not url or url in seen_urls:
            continue
        seen_urls.add(url)

        published = item.get("publishedDate")
        if published:
            try:
                from datetime import datetime
                published = datetime.fromisoformat(published.replace("Z", "+00:00"))
                published = published.strftime("%Y-%m-%d")
            except (ValueError, AttributeError):
                published = None

        parsed.append({
            "title": item.get("title", "Untitled"),
            "url": url,
            "snippet": content[:500] if content else "",
            "engine": item.get("engine", "unknown"),
            "publishedDate": published
        })

    return parsed